packages/contracts/evm-contracts/contracts/token/InverseAppProjected1155.sol

```diff
@@ -0,0 +1,173 @@
+
+struct MintEntry {
+    address minter;
+    uint256 userTokenId;
+}
+
+/// @dev A Paima Inverse Projection ERC1155 token where initialization is handled by the app-layer.
+/// A standard ERC1155 that can be freely minted and stores an unique `<minter, userTokenId>` pair (used in `tokenURI`) when minted.
+contract InverseAppProjected1155 is IInverseAppProjected1155, ERC1155Supply, Ownable {
+    using Strings for uint256;
+
+    string public name;
+    string public symbol;
+
+    mapping(uint256 id => MintEntry) public tokenToMint;
+    mapping(address minter => uint256) public mintCount;
+    mapping(uint256 id => uint256) private _initialSupply;
+
+    /// @dev The token ID that will be minted when calling the `mint` function.
+    uint256 public currentTokenId;
+    /// @dev Base URI that is used in the `uri` function to form the start of the token URI.
+    string public baseURI;
+    /// @dev Base extension that is used in the `uri` function to form the end of the token URI.
+    string public baseExtension;
+
+    /// @dev Sets the token's `_name`, `_symbol`, and transfers ownership to `_owner`.
+    /// Also sets `currentTokenId` to 1.
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _owner
+    ) ERC1155("") Ownable(_owner) {
+        name = _name;
+        symbol = _symbol;
+        currentTokenId = 1;
+    }
+
+    /// @dev Returns true if this contract implements the interface defined by `interfaceId`. See EIP165.
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override(IERC165, ERC1155) returns (bool) {
+        return
+            interfaceId == type(IInverseProjected1155).interfaceId ||
+            interfaceId == type(IInverseAppProjected1155).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    function currentNonce(address user) external view virtual returns (uint256) {
+        return mintCount[user];
+    }
+
+    function validateMint(address, bytes memory) internal virtual returns (bool) {
+        // Base contract allows any mint
+        // Replace this with any custom verification logic
+        return true;
+    }
+
+    /// @dev Returns the amount of token with ID `id` that had been initially minted.
+    function initialSupply(uint256 id) public view virtual returns (uint256) {
+        return _initialSupply[id];
+    }
+
+    /// @dev Mints `value` of a new token to the transaction sender.
+    /// Increases the `currentTokenId`.
+    /// Reverts if transaction sender is a smart contract that does not implement IERC1155Receiver-onERC1155Received.
+    /// Emits the `Minted` event.
+    /// @param value the amount of tokens to mint.
+    /// @param data additional data to pass to the receiver contract.
+    /// @param verificationData any additional data to verify the validity of the mint
+    /// @return id of the minted token.
+    function mint(
```

Comment:
I noticed the mint here is virtual unlike the inverse projected ERC721. Do we want to unify this?

Reply:
Yes! I realized when implementing this that we should basically make all the external functions in the standard contracts virtual, to allow users to override it. I will make that unification in the next PR.